documentation/src/main/asciidoc/user_guide/chapter-63-Streams.adoc

```diff
@@ -0,0 +1,416 @@
+
+=== Common stream operations
+
+This section highlights various options that are present irrespective of what type of underlying cache
+you are using.
+
+==== Key filtering
+
+It is possible to filter the stream so that it only operates upon a given subset of keys.  This can be done
+by invoking the
+link:https://docs.jboss.org/infinispan/{infinispanversion}/apidocs/org/infinispan/CacheStream.html#filterKeys-java.util.Set-[filterKeys]
+method on the `CacheStream`.  This should always be used over a manual filter and will be faster.  Also this
+method allows for distributed computations unlike if you manually retrieve these entries using
+link:https://docs.jboss.org/infinispan/{infinispanversion}/apidocs/org/infinispan/AdvancedCache.html#getAll-java.util.Set-[getAll]
+and you would have to process locally (pick the best method for the job).
+
+==== Segment based filtering
+
+NOTE: This is an advanced feature and should only be used with deep knowledge of Infinispan segment and hashing techniques.
+
+This option is only supported for replicated and distributed caches.  This allows the user to operate upon
+a subset of data at a time as determined by the
+link:https://docs.jboss.org/infinispan/{infinispanversion}/apidocs/org/infinispan/distribution/DistributionManager.html[DistributionManager].
+The segments can be filtered by invoking
+link:https://docs.jboss.org/infinispan/{infinispanversion}/apidocs/org/infinispan/CacheStream.html#filterKeySegments-java.util.Set-[filterKeySegments]
+method on the `CacheStream`.  This is applied after the key filter above but before any intermediate operations are performed.
+
+=== Local/Invalidation/Replication Cache
+
+A stream used with a local, invalidation or replication cache can be used just the same way you would use a stream on a
+regular collection. Infinispan handles all of the translations if necessary behind the scenes and works with all
+of the more interesting options (ie. storeAsBinary, compatibility mode, and a cache loader).
+
+An example of a simple stream task is as follows
+
+==== Example
+
+Below takes a given map and assuming the values are Strings it filters all of the entries that contain the
+word JBoss and then creates a new Map with only the results of the entries that contained JBoss as a value.
+[source,java]
+----
+Map<Object, String> jbossValues = cache.entrySet().stream()
+              .filter(e -> e.getValue().contains("Jboss"))
+              .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+----
+
+=== Distribution Cache
+
+This is where streams come into their stride.  When a stream operation is performed it will send the various intermediate and
+terminal operations to each node that has pertinent data.  This allows the resulting data to be sent back to the
+originating node and intermediate values are not required to be returned which results in much better performance.
```

Comment:
This sounds a bit cleaner to me: "This allows processing the intermediate values on the nodes owning the data, and only sending the final results back to the originating nodes, improving performance."